src/backend/distributed/planner/deparse_shard_query.c

```diff
@@ -0,0 +1,238 @@
+			Query *copiedSubquery = NULL;
+			List *relationShardList = task->relationShardList;
+			ShardInterval *shardInterval = LoadShardInterval(task->anchorShardId);
+
+			query = copyObject(originalQuery);
+
+			copiedInsertRte = ExtractInsertRangeTableEntry(query);
+			copiedSubqueryRte = ExtractSelectRangeTableEntry(query);
+			copiedSubquery = copiedSubqueryRte->subquery;
+
+			AddShardIntervalRestrictionToSelect(copiedSubquery, shardInterval);
+			ReorderInsertSelectTargetLists(query, copiedInsertRte, copiedSubqueryRte);
+
+			/* setting an alias simplifies deparsing of RETURNING */
+			if (copiedInsertRte->alias == NULL)
+			{
+				Alias *alias = makeAlias(CITUS_TABLE_ALIAS, NIL);
+				copiedInsertRte->alias = alias;
+			}
+
+			UpdateRelationToShardNames((Node *) copiedSubquery, relationShardList);
+		}
+
+		deparse_shard_query(query, relationId, task->anchorShardId,
+							newQueryString);
+
+		ereport(DEBUG4, (errmsg("query before rebuilding: %s",
+								task->queryString)));
+		ereport(DEBUG4, (errmsg("query after rebuilding:  %s",
+								newQueryString->data)));
+
+		task->queryString = newQueryString->data;
+	}
+}
+
+
+/*
+ * UpdateRelationToShardNames walks over the query tree and appends shard ids to
+ * relations. It uses unique identity value to establish connection between a
+ * shard and the range table entry. If the range table id is not given a
+ * identity, than the relation is not referenced from the query, no connection
+ * could be found between a shard and this relation. Therefore relation is replaced
+ * by set of NULL values so that the query would work at worker without any problems.
+ *
+ */
+bool
+UpdateRelationToShardNames(Node *node, List *relationShardList)
+{
+	RangeTblEntry *newRte = NULL;
+	uint64 shardId = INVALID_SHARD_ID;
+	Oid relationId = InvalidOid;
+	Oid schemaId = InvalidOid;
+	char *relationName = NULL;
+	char *schemaName = NULL;
+	bool replaceRteWithNullValues = false;
+	ListCell *relationShardCell = NULL;
+	RelationShard *relationShard = NULL;
+
+	if (node == NULL)
+	{
+		return false;
+	}
+
+	/* want to look at all RTEs, even in subqueries, CTEs and such */
+	if (IsA(node, Query))
+	{
+		return query_tree_walker((Query *) node, UpdateRelationToShardNames,
+								 relationShardList, QTW_EXAMINE_RTES);
+	}
+
+	if (!IsA(node, RangeTblEntry))
+	{
+		return expression_tree_walker(node, UpdateRelationToShardNames,
+									  relationShardList);
+	}
+
+	newRte = (RangeTblEntry *) node;
+
+	if (newRte->rtekind != RTE_RELATION)
+	{
+		return false;
+	}
+
+	/*
+	 * Search for the restrictions associated with the RTE. There better be
+	 * some, otherwise this query wouldn't be elegible as a router query.
+	 *
+	 * FIXME: We should probably use a hashtable here, to do efficient
+	 * lookup.
+	 */
+	foreach(relationShardCell, relationShardList)
+	{
+		relationShard = (RelationShard *) lfirst(relationShardCell);
+
+		if (newRte->relid == relationShard->relationId)
```

Comment:
I am finding this difficult to understand. The previous GetRteIdentity call was used to differentiate between multiple usages of the same relation id. Especially when the same relation was used in different levels of nested subqueries.

We have given each relation RTE a unique Id in the beginning and search for that particular id since relation Id by itself was not enough.